zeus/carbon/carbon.py

```diff
@@ -0,0 +1,95 @@
+"""Carbon intensity providers used for carbon-aware optimizers."""
+
+from __future__ import annotations
+
+import abc
+import json
+import logging
+import requests
+
+logger = logging.getLogger(__name__)
+
+
+def get_ip_lat_long() -> tuple[float, float]:
+    """Retrieve the latitude and longitude of the current IP position."""
+    try:
+        ip_url = "http://ipinfo.io/json"
+        resp = requests.get(ip_url)
+        loc = resp.json()["loc"]
+        lat, long = map(float, loc.split(","))
+        logger.info("Retrieved latitude and longitude: %s, %s", lat, long)
+        return lat, long
+    except requests.exceptions.RequestException as e:
+        logger.exception(
+            "Failed to retrieve current latitude and longitude of IP: %s", e
+        )
+        raise
+
+
+class CarbonIntensityNotFoundError(Exception):
+    """Exception when carbon intensity measurement could not be retrieved."""
+
+    def __init__(self, message: str) -> None:
+        """Initialize carbon not found exception."""
+        super().__init__(message)
+
+
+class CarbonIntensityProvider(abc.ABC):
+    """Abstract class for implementing ways to fetch carbon intensity."""
+
+    def __init__(
+        self,
+        location: tuple[float, float],
+        estimate: bool = False,
+        emission_factor_type: str = "direct",
+    ) -> None:
+        """Initializes carbon intensity provider location to the latitude and longitude of the input `location`.
+
+        Args:
+            location: tuple of latitude and longitude (latitude, longitude)
+            estimate: bool to toggle whether carbon intensity is estimated or not
+            emission_factor_type: emission factor to be measured (`direct` or `lifestyle`)
+        """
+        self.lat, self.long = location
+        self.estimate = estimate
+        self.emission_factor_type = emission_factor_type
```

Comment:
`estimate` and `emission_factor_type` are specific to ElectricityMaps, so they shouldn't be here. Move the whole `__init__` definition to `ElectricityMapsClient`. `CarbonIntensityProvider` doesn't have to define a constructor.

Reply:
Also, the type of `emission_factor_type` should be `Literal["direct", "lifecycle"]`. Ref: https://docs.python.org/3/library/typing.html#typing.Literal

Reply:
Gotcha. Would it be better if I leave the abstract class constructor to hold and initialize`lat` and `long`, and then the derived class initializes `emission_factor_type` and `estimate` and calls the `super().__init__(location)`?

Reply:
Nah, I think it's better to leave how `__init__` implemented entirely to the child class. If the parent class has an `__init__`, child classes are forced to respect it.